Framework/Algorithms/src/MaskDetectorsIf.cpp

```diff
@@ -15,123 +16,148 @@ DECLARE_ALGORITHM(MaskDetectorsIf)
+}
+
+/**
+ * Validates the algorithm's input properties.
+ * @return A map from property name to reported issue
+ */
+std::map<std::string, std::string> MaskDetectorsIf::validateInputs() {
+  std::map<std::string, std::string> issues;
+  const auto noInputFile = isDefault("InputCalFile");
+  const auto noOutputFile = isDefault("OutputCalFile");
+  if (!noInputFile && noOutputFile) {
+    issues["OutputCalFile"] = "Output file name is missing.";
+  } else if (noInputFile && !noOutputFile) {
+    issues["InputCalFile"] = "Input file name is missing.";
+  }
+  return issues;
 }
 
 /** Executes the algorithm
- *
- *  @throw Exception::FileError If the grouping file cannot be opened or read
- *successfully
- *  @throw std::runtime_error If the rebinning process fails
  */
 void MaskDetectorsIf::exec() {
   retrieveProperties();
-  const size_t nspec = inputW->getNumberHistograms();
 
+  if (isDefault("InputCalFile") && isDefault("OutputWorkspace")) {
+    g_log.error() << "No InputCalFle or OutputWorkspace specified; the "
+                     "algorithm will do nothing.";
+    return;
+  }
+  const size_t nspec = m_inputW->getNumberHistograms();
   for (size_t i = 0; i < nspec; ++i) {
     // Get the list of udets contributing to this spectra
-    const auto &dets = inputW->getSpectrum(i).getDetectorIDs();
+    const auto &dets = m_inputW->getSpectrum(i).getDetectorIDs();
 
     if (dets.empty())
       continue;
     else {
-      double val = inputW->y(i)[0];
-      if (compar_f(val, value)) {
-        for (auto det : dets) {
-          umap.emplace(det, select_on);
+      const double val = m_inputW->y(i)[0];
+      if (m_compar_f(val, m_value)) {
+        for (const auto det : dets) {
+          m_umap.emplace(det, m_select_on);
         }
       }
     }
-    double p = static_cast<double>(i) / static_cast<double>(nspec);
+    const double p = static_cast<double>(i) / static_cast<double>(nspec);
     progress(p, "Generating detector map");
   }
-  std::string oldf = getProperty("InputCalFile");
-  std::string newf = getProperty("OutputCalFile");
-  progress(0.99, "Creating new cal file");
-  createNewCalFile(oldf, newf);
+
+  if (!isDefault("InputCalFile")) {
+    createNewCalFile();
+  }
+  if (!isDefault("OutputWorkspace")) {
+    outputToWorkspace();
+  }
+}
+
+/**
+ * Create an output workspace masking/unmasking the selected/deselected spectra
+ */
+void MaskDetectorsIf::outputToWorkspace() {
+  API::MatrixWorkspace_sptr outputW = getProperty("OutputWorkspace");
+  if (outputW != m_inputW)
+    outputW = m_inputW->clone();
+  auto &detectorInfo = outputW->mutableDetectorInfo();
+  const size_t nspec = outputW->getNumberHistograms();
+  for (auto it = m_umap.cbegin(); it != m_umap.cend(); ++it) {
+    detectorInfo.setMasked(detectorInfo.indexOf(it->first), it->second);
+    const double p =
+        static_cast<double>(std::distance(m_umap.cbegin(), it) + 1) /
+        static_cast<double>(nspec);
+    progress(p, "Generating detector map");
```

Comment:
I think you are right. Further, the call to `progress()` here is actually a leftover from refactoring I did after the initial prototype so I will remove it. Maybe I should reconsider the progress reporting also in the loop in `exec()` as well since it has the same problem.